File: src/network/packet/impl/PlayerDisconnectPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl;

use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\packet\CloudboundPacket;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\util\PacketData;
use pocketcloud\cloud\player\CloudPlayerManager;

final class PlayerDisconnectPacket extends CloudPacket implements CloudboundPacket {

    public function __construct(private ?string $player = null) {}

    public function handle(ServerClient $client): void {
        if (($player = CloudPlayerManager::getInstance()->get($this->player)) !== null) {
            if ($player->getCurrentProxy() === null) {
                CloudPlayerManager::getInstance()->remove($player);
            } else {
                if (($server = $client->getServer()) !== null) {
                    if ($server->getTemplate()->getTemplateType()->isProxy()) {
                        CloudPlayerManager::getInstance()->remove($player);
                    }
                }
            }
        }
    }

    public function encodePayload(PacketData $packetData): void {}

    public function decodePayload(PacketData $packetData): void {
        $packetData->readAll($this->player);
    }

    public function getPlayer(): ?string {
        return $this->player;
    }

    public static function create(string $player): self {
        return new self($player);
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/impl/PlayerDisconnectPacket.java
package de.pocketcloud.cloud.network.packet.impl;

import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.packet.CloudboundPacket;
import de.pocketcloud.cloud.network.packet.CloudPacket;
import de.pocketcloud.cloud.network.packet.data.PacketData;
import de.pocketcloud.cloud.player.CloudPlayerManager;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;

@NoArgsConstructor
@Getter
public final class PlayerDisconnectPacket extends CloudPacket implements CloudboundPacket {

    private String player;

    public PlayerDisconnectPacket(String player) {
        this.player = player;
    }

    @Override
    public void handle(@NotNull ServerClient client) {
        var cloudPlayer = CloudPlayerManager.instance().get(player).orElse(null);
        if (cloudPlayer != null) {
            if (cloudPlayer.currentProxy().isEmpty()) {
                CloudPlayerManager.instance().remove(cloudPlayer);
            } else {
                var server = client.server();
                if (server != null && server.template().templateType().isProxy()) {
                    CloudPlayerManager.instance().remove(cloudPlayer);
                }
            }
        }
    }

    @Override
    public void encodePayload(PacketData packetData) {}

    @Override
    public void decodePayload(PacketData packetData) {
        this.player = packetData.readString();
    }

    public static PlayerDisconnectPacket create(String player) {
        return new PlayerDisconnectPacket(player);
    }
}
